libcudacxx/include/cuda/__functional/maximum.h

```diff
@@ -0,0 +1,61 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_FUNCTIONAL_MAXIMUM_H
+#define _CUDA_FUNCTIONAL_MAXIMUM_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__algorithm/max.h>
+#include <cuda/std/__functional/binary_function.h>
+#include <cuda/std/__utility/forward.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+template <class _Tp = void>
+struct _CCCL_TYPE_VISIBILITY_DEFAULT maximum : __binary_function<_Tp, _Tp, _Tp>
+{
+  using __result_type = _Tp; // used by valarray
+
+  _CCCL_EXEC_CHECK_DISABLE
+  constexpr _LIBCUDACXX_HIDE_FROM_ABI _Tp operator()(const _Tp& __x, const _Tp& __y) const
```

Comment:
- `std::max<T>` returns a `const T&`.
- `cuda::maximum<T>::operator()` returns a `T`.
- `cuda::maximum<void>::operator()<T,T>` return a `const T&`.

That's a gratuitous difference that will cause problems eventually.  Either:
- `cuda::maximum<T>::operator()` should return a `const T&`.  This seems like the correct choice because it matches the return type of `std::max<T>`, and because it can be used with non-copyable types.  But it results in a slight change to `thrust::maximum<T>::operator()`, which could in theory have cascading effects.
- `cuda::maximum<void>::operator()<T,T>` should return a `T` so it is consistent with the non-`void` specializations.

I'm not sure which is the better choice.

`cuda::minimum` has the same issue.

Reply:
Hmmm I think returning `T` is the best choice. Consider the following:

1. All of the other functor objects (e.g. `std::plus<>`, `std::minus<>` etc.) return by value, so having `minimum` and `maximum` return by ref would be inconsistent.
2. It just so happens that we implement using `cuda::std::min()`. I could just as well have inlined this (and in fact, most host compiler runtimes do).
3. The functor objects are often used as part of reductions. If we return by ref, then you can easily get into subtle bug territory. Consider:
```c++
int x = 0, y = 100;
// auto&& is common in generic code, min_val = x = 0 now
auto&& min_val = cuda::minimum<>{}(x, y);
// oops, min_val = 1 now too
++x;
```
Sure you could argue that this is a programmer mistake (using `auto&&` when it should just be `auto`, and take a copy). But if we return by-value, then this mistake is impossible.

If you agree, I will update `cuda::maximum<void>::operator()` to return a `T`. I guess a follow-on question should be:
1. Take only 1 template param?
2. Or take 2 template param and return `std::common_type_t`?

Reply:
> All of the other functor objects (e.g. std::plus<>, std::minus<> etc.) return by value

That's a compelling argument.  I should have checked that myself.  So yes, I agree that `maximum<void>` should be changed to return `T`.

I prefer two template parameters and returning `std::common_type_t`.  And have the implementation be a `?:` expression rather than a call to `cuda::std::max`.

Reply:
OK return `common_type` and inlined `max()` and `min()`. Let me know if I missed anything in the conversion. STL code is quite... hard on the eyes :)

Reply:
For completeness: `cuda::std::identity` returns `T&&` and `thrust::identity` returns `const T&`, so returning a `T` for `maximum/minimum` is not a must.

Reply:
Conventionally attributes should come first
```suggestion
  _LIBCUDACXX_HIDE_FROM_ABI constexpr _Tp operator()(const _Tp& __x, const _Tp& __y) const
```